doc/kernel/usermode/kernelobjects.rst

```diff
@@ -0,0 +1,223 @@
+  may grant permission on an object to another thread via the
+  :c:func:`k_object_access_grant()` API. The convenience function
+  :c:func:`k_thread_access_grant()` may also be used, which accepts a
+  NULL-terminated list of kernel objects and calls
+  :c:func:`k_object_access_grant()` on each of them. The thread being granted
+  permission must already be initialized, but the objects do not need to be.
+  If the caller is from user mode, the caller must have permissions on both
+  the kernel object and the target thread object.
+
+* Supervisor threads may declare a particular kernel object to be a public
+  object, usable by all current and future threads with the
+  :c:func:`k_object_access_all_grant()` API. You must assume that any
+  untrusted or exploited code will then be able to access the object, use
+  this API with caution.
+
+* If a thread was declared statically with :c:macro:`K_THREAD_DEFINE()`,
+  then the :c:macro:`K_THREAD_ACCESS_GRANT()` may be used to grant that thread
+  access to a set of kernel objects at boot time.
+
+Once a thread has been granted access to an object, such access may be
+removed with the :c:func:`k_object_access_revoke()` API. User threads using
+this API must have permission on both the object in question, and the thread
+object that is having access revoked.
+
+Initialization State
+====================
+
+Most operations on kernel objects will fail if the object is considered to be
+in an unitialized state. The appropriate init function for the object must
+be performed first.
+
+Some objects will be implicitly initialized at boot:
+
+* Kernel objects that were declared with static initialization macros
+  (such as :c:macro:`K_SEM_DEFINE` for semaphores) will be in an initialized
+  state at build time.
+
+* Device driver objects are considered initilaized after their init function
+  is run by the kernel early in the boot process.
+
+If a kernel object is intitialized with a private static initailizer, the
+object must have :c:func:`_k_object_init()` on it at some point by a supervisor
+thread, otherwise the kernel will consider the object uninitialized if accessed
+by a user thread. This is very uncommon, typically only for kernel objects that
+are embedded within some larger struct and initialized statically.
+
+.. code-block:: c
+
+    struct foo {
+        struct k_sem sem;
+        ...
+    };
+
+    __kernel struct foo my_foo = {
+        .sem = _K_SEM_INITIALIZER(my_foo.sem, 0, 1),
+        ...
+    };
+
+    ...
+    _k_object_init(&my_foo.sem);
+    ...
+
+
+Creating New Kernel Object Types
+================================
+
+When implementing new kernel features or driver subsystems, it may be necessary
+to define some new kernel object types. There are different steps needed for
+core kernel objects and new driver subsystems.
+
+Creating New Core Kernel Objects
+--------------------------------
+
+* In scripts/gen_kobject_list.py, add the name of the struct to the
```

Comment:
nice to mark scripts with double back-ticks:

     ``scripts/gen_kobject_list.py``